### PR TITLE
feat(APIv2): RHICOMPL-3786 implement Security Guides controller

### PR DESCRIPTION
**Pull request by xmicha82**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: spec/controllers/v2/security_guides_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+          get :index
+
+          expected = {
+            'id': sg.id,
+            'ref_id': sg.ref_id,
+            'version': sg.version
+          }
+
+          expect(response.parsed_body).to include(expected)
```
Comment (skateman) [inline spec/controllers/v2/security_guides_controller_spec.rb:40]: You can merge this and the previous test together, meaning the response expectation can be moved here and the previous one can be deleted. Then the test itself talks about security guides in plural but you are just testing the existence of a single entity in it, not even the fact if it is an array or not. Furthermore, you can write some smart iteration to check all 3 field in a one-liner if really necessary...

File: spec/controllers/v2/security_guides_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+        it 'paginates' do
+          PER_PAGE = 10
+
+          get :index, params: { 'limit' => PER_PAGE }
+
+          binding.pry
+
+          expect(response.parsed_body['data'].count).to eq(PER_PAGE)
+        end
```
Comment (skateman) [inline spec/controllers/v2/security_guides_controller_spec.rb:51]: Do you really test pagination here? It seems to me that you're just checking if the limit parameter works properly.

File: config/routes.rb [hunk 3]
```diff
@@ -22,6 +22,16 @@ def draw_routes(prefix)
         end
       end
 
+      unless Rails.env.production?
+        scope 'v2', module: 'v2', as: 'v2' do
+          resources :security_guides, only: [:index, :show] do
+            resources :profiles, only: [:index, :show]
+            resources :rules, only: [:index, :show]
+            resources :value_definitions, only: :index
```
Comment (skateman) [inline config/routes.rb:30]: How these routes are part of the scope? I think they should be removed from this PR.

File: app/controllers/v2/application_controller.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    class << self
+      def permission_for_action(action, permission)
+        @action_permissions ||= {}
+        @action_permissions[action.to_sym] ||= permission
+      end
+    end
+
+    def openapi
+      raise NotImplementedError # TODO
+    end
```
Comment (skateman) [inline app/controllers/v2/application_controller.rb:28]: This is no longer needed, we serve the JSON as a static route.

File: app/serializers/v2/application_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    #   # rubocop:disable Layout/LineLength
+    #   def relationships_hash(record, relationships, fieldset, includes_list, params = {})
+    #     # rubocop:enable Layout/LineLength
+    #     relationships_disabled = params[:root_resource] != record.class ||
+    #                              !params[:relationships]
+    #     return {} if relationships_disabled
+
+    #     super(record, relationships, fieldset, includes_list, params)
+    #   end
+    # end
```
Comment (skateman) [inline app/serializers/v2/application_serializer.rb:20]: Please drop this, probably not needed at all.

File: spec/controllers/v2/security_guides_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
```
Comment (skateman) [inline spec/controllers/v2/security_guides_controller_spec.rb:3]: From where this name is coming from? Shouldn't this be `spec_helper`? Would it work without it?

**Review by skateman — CHANGES_REQUESTED** [review]

File: spec/support/shared_examples/with_metadata_spec.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+RSpec.shared_examples 'with metadata' do
+  let(:item_count) { 20 }
+  let(:meta_keys) { %w[total limit offset relationships] }
+
+  it 'contains total, limit, offset and relationships keys' do
```
Comment (skateman) [inline spec/support/shared_examples/with_metadata_spec.rb:7]: Yes, it contains, but what abouts its content(s)? You are only testing the `item_count` below...but the rest is untested :disappointed:

File: config/environments/test.rb [hunk 8]
```diff
@@ -63,5 +63,5 @@
   # Disable logging
   config.logger = ActiveSupport::TaggedLogging.new(Insights::Api::Common::LoggerWithAudit.new(File::NULL))
 
-  config.factory_bot.definition_file_paths = ["test/factories"]
+  config.factory_bot.definition_file_paths = ["test/factories", "spec/factories"]
```
Comment (romanblanco) [inline config/environments/test.rb:66]: The `spec/factories` here will eventually (in follow-up PRs) replace the `test/factories`.

Am I understanding it right, @xmicha82?

Comment (skateman) [reply config/environments/test.rb:66]: Yes!

File: spec/controllers/v2/sort/security_guides_controller.yaml [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    :result: [[1, 3, 5], 0, 2, 4]
+  - :sort_by:
+      - 'title:asc'
+    :result: [[1, 3, 5], 0, 2, 4]
+  - :sort_by:
+      - 'title:desc'
+    :result: [4, 2, 0, [1, 3, 5]]
+  - :sort_by:
+      - 'version'
+    :result: [1, [5, 3], 4, 2, 0]
```
Comment (romanblanco) [inline spec/controllers/v2/sort/security_guides_controller.yaml:38]: @xmicha82 Why is the `[5, 3]` nested here? Is this case documented somewhere?

**Comment by vkrizan** [comment]
Hey @xmicha82 and @skateman, do you have any documents that explain the efforts and goals of API v2?

File: spec/controllers/v2/sort/security_guides_controller.yaml (hunk 9 above)
Comment (xmicha82) [reply spec/controllers/v2/sort/security_guides_controller.yaml:38]: security guide at indexes 3 and 5 have the same version, so they need to be sorted by ID, which is achieved by nesting them, as explained in `spec/support/shared_exaples/sortable_spec.rb:22`

File: config/environments/test.rb (hunk 8 above)
Comment (xmicha82) [reply config/environments/test.rb:66]: well, `test/factories` are for APIv1 tests, aren't they? so i don't think they will be replaced

File: spec/controllers/v2/sort/security_guides_controller.yaml (hunk 9 above)
Comment (romanblanco) [reply spec/controllers/v2/sort/security_guides_controller.yaml:38]: Could you add some demonstration in the comments of how this nesting/sorting works?

After reading `spec/support/shared_exaples/sortable_spec.rb:22`, the process still feels magical.

File: spec/controllers/v2/security_guides_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+          expect(sg['attributes'].keys.count).to eq(attributes.keys.count)
+        end
+      end
+
+      it_behaves_like 'searchable'
+      it_behaves_like 'paginable'
+      it_behaves_like 'sortable'
+      include_examples 'with metadata'
+
+      context 'RBAC denied' do
```
Comment (romanblanco) [inline spec/controllers/v2/security_guides_controller_spec.rb:55]: @xmicha82 this should not be in 'Authorized' context.

File: spec/controllers/v2/security_guides_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+          end
+        end
+
+        it_behaves_like 'searchable'
+        it_behaves_like 'paginable'
+        it_behaves_like 'sortable'
+        include_examples 'with metadata'
+      end
+
+      context 'RBAC denied' do
```
Comment (romanblanco) [inline spec/controllers/v2/security_guides_controller_spec.rb:58]:
```suggestion
      context 'Unauthorized' do
```

File: spec/controllers/v2/security_guides_controller_spec.rb [hunk 12]
```diff
@@ -26,35 +25,38 @@
...
-      it_behaves_like 'sortable'
-      include_examples 'with metadata'
+        it_behaves_like 'searchable'
+        it_behaves_like 'paginable'
+        it_behaves_like 'sortable'
+        include_examples 'with metadata'
+      end
 
-      context 'RBAC denied' do
+      context 'Unathorized' do
```
Comment (skateman) [inline spec/controllers/v2/security_guides_controller_spec.rb:58]: These changes aren't in the scope of the commit message

**Review by skateman — APPROVED** [review]
The Seal of Approval

**Pull request merged**